### PR TITLE
Make all proxies, even gateways, have SidecarScope

### PR DESCRIPTION
**Pull request by howardjohn**
By giving gateways a `egress: */*` we can simplify the code and use the
assumption that sidecarscope is never nil. This is a stepping stone to
removing the old EDS legacy code entirely

This code is essentially a subset of https://github.com/istio/istio/pull/15531. That PR tries to do two things at once which makes it hard to understand and debug.

[x] Networking

### CODE REVIEW

**Comment by howardjohn** [comment]
/hold

I want to make sure this doesn't negatively impact performance before merging

File: pilot/pkg/model/context.go [hunk 1]
```diff
@@ -205,8 +205,14 @@ func (node *Proxy) GetRouterMode() RouterMode {
 // Listener generation code will still use the SidecarScope object directly
 // as it needs the set of services for each listener port.
 func (node *Proxy) SetSidecarScope(ps *PushContext) {
-	labels := node.WorkloadLabels
-	node.SidecarScope = ps.getSidecarScope(node, labels)
+	if node.Type == SidecarProxy {
+		labels := node.WorkloadLabels
+		node.SidecarScope = ps.getSidecarScope(node, labels)
+	} else {
+		// Gateways should just have a default scope with egress: */*
```
Comment (costinm) [inline pilot/pkg/model/context.go]: Not sure - Gateway has now a way to specify an explicit set of namespaces. 
Maybe ok for now ( not changing behavior ), but we should add a TODO: to scan if the Gateway 
is using only "namespace/foo" for redirection that this is used as equivalent with the Sidecar, i.e. 
ignore any namespace not explicitly referenced by Gateway.

File: pilot/pkg/networking/core/v1alpha3/listener.go [hunk 2]
```diff
@@ -705,208 +704,172 @@ func (configgen *ConfigGeneratorImpl) buildSidecarOutboundListeners(env *model.E
...
 						proxyInstances: proxyInstances,
 						proxyLabels:    proxyLabels,
+						port:           servicePort.Port,
 						bind:           bind,
-						port:           listenPort.Port,
 						bindToPort:     bindToPort,
 					}
 
 					pluginParams := &plugin.InputParams{
-						ListenerProtocol:           plugin.ModelProtocolToListenerProtocol(listenPort.Protocol),
```
Comment (costinm) [inline pilot/pkg/networking/core/v1alpha3/listener.go]: Would be great to not mix renames with larger changes..

File: pilot/pkg/networking/core/v1alpha3/listener.go [hunk 3]
```diff
@@ -705,208 +704,172 @@ func (configgen *ConfigGeneratorImpl) buildSidecarOutboundListeners(env *model.E
...
+					Port:                       listenPort,
 					Service:                    service,
 				}
 
-				configgen.buildSidecarOutboundListenerForPortOrUDS(listenerOpts, pluginParams,
-					listenerMap, virtualServices, actualWildcard)
+				configgen.buildSidecarOutboundListenerForPortOrUDS(listenerOpts, pluginParams, listenerMap,
+					virtualServices, actualWildcard)
 			}
-		}
```
Comment (costinm) [inline pilot/pkg/networking/core/v1alpha3/listener.go]: Can you explain a bit more why this change was made ? Is it related to 'having a default sidecar', or just a separate change ?

File: pilot/pkg/model/context.go (hunk 1 above)
Comment (howardjohn) [reply pilot/pkg/model/context.go]: @mandarjog had a PR a while back about scoping gateways. Not sure what happened to it. Maybe this would be relevant for that

File: pilot/pkg/networking/core/v1alpha3/listener.go (hunk 3 above)
Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/listener.go]: This change is a weird diff from being unindented. Before we had a special case of sidecarScope == nil, but now we have a default scope for sidecar proxies and gateways have a sidecar scope, so we no longer need the special case

File: pilot/pkg/proxy/envoy/v2/eds.go [hunk 4]
```diff
@@ -692,13 +691,7 @@ func (s *DiscoveryServer) pushEds(push *model.PushContext, con *XdsConnection, v
 			}
 		}
 
-		var l *xdsapi.ClusterLoadAssignment
-		// decide which to use based on presence of Sidecar.
-		if sidecarScope == nil || sidecarScope.Config == nil || len(features.DisableEDSIsolation) != 0 {
```
Comment (costinm) [inline pilot/pkg/proxy/envoy/v2/eds.go]: So we're also removing the 'disable EDS isolation' ? I guess it's ok.

Comment (howardjohn) [reply pilot/pkg/proxy/envoy/v2/eds.go]: I guess this doesn't need to be in this change. This change was broken off the larger change of completely removing the legacy EDS code, but that turned out to be a lot more complicated. Should be no problems keeping it if we still want it, although it seems safe to remove now
